### ISSUE TITLE
feature request: config notebook with specified configuration files

### ISSUE THREAD

**Issue by shelper** [jupyter/notebook#2425]
currently the configuration file is` jupyter_notebook_config.py `in the folder `~/.jupyter`. 

i would suggest to provide the freedom of specifying which config file to load when starting Jupiter notebook

for instance, i would like to setup extensions for a project in a specific folder, i can do:
```sh
jupyter notebook --config ./config_for_this_folder.py
```
this will help the Jupiter notebook configuration information to be version controllable and explicable on different PCs / platforms

**Comment by takluyver** [comment]
`--config PATH` should already work. Have you tried it?

**Comment by shelper** [comment]
hummm, why i didnt find this information in the documentation?

anyway, it worked, thanks!

**Comment by shelper** [comment]
tried the `--config=` option, and it didnt work,... my environment:

Python	3.5.2 64bit [MSC v.1900 64 bit (AMD64)]
IPython	5.1.0
OS	Windows 7 6.1.7601 SP1

**Comment by takluyver** [comment]
What happens when you try it? I've just checked and it's working for me.

**Comment by shelper** [comment]
so, even  I put something like :
`jupyter notebook --config="non_existing_file.py"`
i don't see any error message in the console window, 
and jupyter notebook just run...

**Comment by takluyver** [comment]
It doesn't fail if the config file is missing, but for me it does use the config file if you pass one that exists.

**Comment by shelper** [comment]
what is the environment you are using? i have the `config.py` in my folder and what i did is:
`jupyter notebook --config="./config.py"`

are you using the exact same format as above?

**Comment by shelper** [comment]
sorry, i think you are right, it is working, i was confused by the config file as thought the config file is something preloaded before you run notebook cells, i put in the config file `import sys` and i thought i should be able to run `sys.path` but i got message saying `sys module is not found`.

so is there a way to run a script file before you run your notebook cells? i know you can put files in `~/.ipython/profile_default/startup/` to be run, but is there a way to specify a script file to run other than files  in this folder?

**Comment by takluyver** [comment]
That's part of the kernel config, rather than the notebook. Have a look for `exec_files` in this document: http://ipython.readthedocs.io/en/stable/config/options/kernel.html

**Comment by shelper** [comment]
still didn't work...
i have `startup.ipy` that contains one line:
```python
import sys
```
then in my `config.py`,  i put 
```python
c.InteractiveShellApp.exec_files = ["./startup.ipy"]
print("startup.ipy loaded")
```
and i run `jupyter notebook --config='config.py'`, and in my console, i saw `startup.ipy loaded` got printed, which means the `config.py` is loaded, but i still got error `NameError: name 'sys' is not defined.`

i also tried to put the lines below into `config.py`:
```python
c.InteractiveShellApp.exec_lines = [ "import sys" ]
print("sys imported")
```
and i saw `sys imported` got printed in the cosole, but i still got error `NameError: name 'sys' is not defined.

**Comment by takluyver** [comment]
Yeah, the `--config` option there only works for the notebook server's config, not the kernel's config. The kernel finds its own configuration separately. There isn't a convenient way to pass kernel config in from the command line.

**Comment by shelper** [comment]
so is there a solution for this? basically i want to run some lines before i run notebook cells when i start notebook, and c.InteractiveShellApp.exec_files does not work....

**Comment by takluyver** [comment]
You can put it in the kernel config file, but there isn't an easy way to specify kernel config when you launch the notebook server.

You could put the code you want to run in a file called e.g. `preamble.py`, and then start each notebook with `import preamble`.
